src/decoder/lossless.rs

```diff
@@ -0,0 +1,212 @@
+                                mcus_left_until_restart = self.restart_interval;
+                            }
+                            Some(marker) => {
+                                return Err(Error::Format(format!(
+                                    "found marker {:?} inside scan where RST{} was expected",
+                                    marker, expected_rst_num
+                                )))
+                            }
+                            None => {
+                                return Err(Error::Format(format!(
+                                    "no marker found where RST{} was expected",
+                                    expected_rst_num
+                                )))
+                            }
+                        }
+                    }
+
+                    mcus_left_until_restart -= 1;
+                }
+
+                for (i, component) in components.iter().enumerate() {
+                    let dc_table = self.dc_huffman_tables[scan.dc_table_indices[i]]
+                        .as_ref()
+                        .unwrap();
+                    let value = huffman.decode(reader, dc_table)?;
+                    let diff = match value {
+                        0 => 0,
+                        1..=15 => huffman.receive_extend(reader, value)? as i32,
+                        16 => 32768,
+                        _ => {
+                            // Section F.1.2.1.1
+                            // Table F.1
+                            return Err(Error::Format(
+                                "invalid DC difference magnitude category".to_owned(),
+                            ));
+                        }
+                    };
+                    if mcu_x > 0 {
+                        ra[i] = results[i][mcu_y * frame.image_size.width as usize + mcu_x - 1];
+                    }
+                    if mcu_y > 0 {
+                        rb[i] = results[i][(mcu_y - 1) * frame.image_size.width as usize + mcu_x];
+                        if mcu_x > 0 {
+                            rc[i] = results[i]
+                                [(mcu_y - 1) * frame.image_size.width as usize + (mcu_x - 1)];
+                        }
+                    }
+                    let prediction = predict(
+                        ra[i] as i32,
+                        rb[i] as i32,
+                        rc[i] as i32,
+                        scan.predictor_selection,
+                        scan.point_transform,
+                        frame.precision,
+                        mcu_x,
+                        mcu_y,
+                        self.restart_interval > 0
+                            && mcus_left_until_restart == self.restart_interval - 1,
+                    );
+                    let result = ((prediction + diff) & 0xFFFF) as u16; // modulo 2^16
+                    results[i].push(result << scan.point_transform);
```

Comment:
This push happens unconditionally, so we are aware of the require memory for each of the result vectors. In other words, they can be pre-allocated with `Vec::with_capacity` before.

Reply:
done